File: packages/node/src/services/account.ts
import { AccountBalanceParams, TransactionStatusParams, ReversalParams } from '../types';
import { HttpClient } from '../http/httpClient';
import { Auth } from '../core/Auth';
import { Helpers } from '../utils/helpers';

export class AccountService {
  private httpClient: HttpClient;
  private auth: Auth;
  private helpers: Helpers;
  private baseUrl: string;

  constructor(httpClient: HttpClient, auth: Auth) {
    this.httpClient = httpClient;
    this.auth = auth;
    this.helpers = new Helpers();
    this.baseUrl = httpClient.getBaseUrl();
  }

  async balance(params: AccountBalanceParams): Promise<any> {
    const { shortcode, identifierType, remarks, resultUrl, timeoutUrl, shortCodeType = 'C2B' } = params;

    const url = `${this.baseUrl}/mpesa/accountbalance/v1/query`;
    const body = {
      Initiator: this.helpers.getConfig('initiatorName'),
      SecurityCredential: this.helpers.generateSecurityCredential(),
      CommandID: 'AccountBalance',
      PartyA: shortcode,
      IdentifierType: identifierType,
      Remarks: remarks,
      ResultURL: resultUrl || this.helpers.getConfig('callbacks.balanceResultUrl'),
      QueueTimeOutURL: timeoutUrl || this.helpers.getConfig('callbacks.balanceTimeoutUrl'),
    };

    const token = await this.auth.getAccessToken(shortCodeType);
    return this.httpClient.post(url, body, token);
  }

  async status(params: TransactionStatusParams): Promise<any> {
    const { shortcode, transactionId, identifierType, remarks, resultUrl, timeoutUrl, shortCodeType = 'C2B' } = params;

    const url = `${this.baseUrl}/mpesa/transactionstatus/v1/query`;
    const body = {
      Initiator: this.helpers.getConfig('initiatorName'),
      SecurityCredential: this.helpers.generateSecurityCredential(),
      CommandID: 'TransactionStatusQuery',
      TransactionID: transactionId,
      PartyA: shortcode,
      IdentifierType: identifierType,
      Remarks: remarks,
      Occassion: '',
      ResultURL: resultUrl || this.helpers.getConfig('callbacks.statusResultUrl'),
      QueueTimeOutURL: timeoutUrl || this.helpers.getConfig('callbacks.statusTimeoutUrl'),
    };

    const token = await this.auth.getAccessToken(shortCodeType);
    return this.httpClient.post(url, body, token);
  }

  async reversal(params: ReversalParams): Promise<any> {
    const { shortcode, transactionId, amount, remarks, resultUrl, timeoutUrl, shortCodeType = 'C2B' } = params;

    const url = `${this.baseUrl}/mpesa/reversal/v1/request`;
    const body = {
      Initiator: this.helpers.getConfig('initiatorName'),
      SecurityCredential: this.helpers.generateSecurityCredential(),
      CommandID: 'TransactionReversal',
      TransactionID: transactionId,
      Amount: amount,
      ReceiverParty: shortcode,
      RecieverIdentifierType: '11',
      Remarks: remarks,
      Occasion: '',
      ResultURL: resultUrl || this.helpers.getConfig('callbacks.reversalResultUrl'),
      QueueTimeOutURL: timeoutUrl || this.helpers.getConfig('callbacks.reversalTimeoutUrl'),
    };

    const token = await this.auth.getAccessToken(shortCodeType);
    return this.httpClient.post(url, body, token);
  }
}

File: packages/python/src/mpesa/services/account.py
from typing import Optional
from .http.http_client import HttpClient
from .auth import Auth
from .utils import Helpers


class AccountService:
    def __init__(self, http_client: HttpClient, auth: Auth, config):
        self.http_client = http_client
        self.auth = auth
        self.helpers = Helpers(config)
        self.base_url = http_client.get_base_url()

    def balance(self, shortcode: str, identifier_type: int, remarks: str,
                result_url: Optional[str] = None, timeout_url: Optional[str] = None,
                short_code_type: str = 'C2B') -> dict:
        url = f'{self.base_url}/mpesa/accountbalance/v1/query'
        
        body = {
            'Initiator': self.helpers.get_config('initiator_name'),
            'SecurityCredential': self.helpers.generate_security_credential(),
            'CommandID': 'AccountBalance',
            'PartyA': shortcode,
            'IdentifierType': identifier_type,
            'Remarks': remarks,
            'ResultURL': result_url or self.helpers.get_config('callbacks.balance_result_url'),
            'QueueTimeOutURL': timeout_url or self.helpers.get_config('callbacks.balance_timeout_url'),
        }
        
        token = self.auth.get_access_token(short_code_type)
        return self.http_client.post(url, body, token)

    def status(self, shortcode: str, transaction_id: str, identifier_type: int, remarks: str,
              result_url: Optional[str] = None, timeout_url: Optional[str] = None,
              short_code_type: str = 'C2B') -> dict:
        url = f'{self.base_url}/mpesa/transactionstatus/v1/query'
        
        body = {
            'Initiator': self.helpers.get_config('initiator_name'),
            'SecurityCredential': self.helpers.generate_security_credential(),
            'CommandID': 'TransactionStatusQuery',
            'TransactionID': transaction_id,
            'PartyA': shortcode,
            'IdentifierType': identifier_type,
            'Remarks': remarks,
            'Occassion': '',
            'ResultURL': result_url or self.helpers.get_config('callbacks.status_result_url'),
            'QueueTimeOutURL': timeout_url or self.helpers.get_config('callbacks.status_timeout_url'),
        }
        
        token = self.auth.get_access_token(short_code_type)
        return self.http_client.post(url, body, token)

    def reversal(self, shortcode: str, transaction_id: str, amount: float, remarks: str,
                 result_url: Optional[str] = None, timeout_url: Optional[str] = None,
                 short_code_type: str = 'C2B') -> dict:
        url = f'{self.base_url}/mpesa/reversal/v1/request'
        
        body = {
            'Initiator': self.helpers.get_config('initiator_name'),
            'SecurityCredential': self.helpers.generate_security_credential(),
            'CommandID': 'TransactionReversal',
            'TransactionID': transaction_id,
            'Amount': amount,
            'ReceiverParty': shortcode,
            'RecieverIdentifierType': '11',
            'Remarks': remarks,
            'Occasion': '',
            'ResultURL': result_url or self.helpers.get_config('callbacks.reversal_result_url'),
            'QueueTimeOutURL': timeout_url or self.helpers.get_config('callbacks.reversal_timeout_url'),
        }
        
        token = self.auth.get_access_token(short_code_type)
        return self.http_client.post(url, body, token)
